packages/auth-providers/clerk/api/src/decoder.ts

```diff
@@ -34,3 +37,39 @@ export const authDecoder: Decoder = async (token: string, type: string) => {
     return Promise.reject(error)
   }
 }
+
+export const jwtAuthDecoder: Decoder = async (token: string, type: string) => {
+  if (type !== 'clerk') {
+    return null
+  }
+
+  const { verifyToken } = await import('@clerk/clerk-sdk-node')
+
+  try {
+    const issuer = (iss: string) =>
+      iss.startsWith('https://clerk.') || iss.includes('.clerk.accounts')
+
+    const jwtPayload = await verifyToken(token, {
+      issuer,
+      apiUrl: process.env.CLERK_API_URL || 'https://api.clerk.dev',
+      jwtKey: process.env.CLERK_JWT_KEY,
+      apiKey: process.env.CLERK_API_KEY,
+      secretKey: process.env.CLERK_SECRET_KEY,
+    })
+
+    if (!jwtPayload.sub) {
+      return Promise.reject(new Error('Session invalid'))
+    }
+
+    // @ts-expect-error need to type this
+    const roles = jwtPayload.sessionClaims.publicMetadata['roles'] ?? []
```

Comment:
Don't let me merge this without fixing this

Reply:
@jtoar I don't think `sessionClaims` exists on `jwtPayload` unless you specifically add it to a jwt template like in the screenshot below:
<img width="610" alt="Screenshot 2023-06-16 at 1 44 26 PM" src="https://github.com/redwoodjs/redwood/assets/1259226/fd742687-a6ac-43fd-8f8b-3a8d3d1bb92f">

Reply:
@o0charlie0o that sounds right, and this is definitely the hardest part of it. The function I used in this PR is from one of the Clerk team members, but I'm not sure what the design considerations were in putting `publicMetadata` on `sessionClaims` vs just listing it as a prop on at the top level like you have here at the bottom.

Given that a user can customize anything about this template, it feels like we either have to 1) recommend a template in the docs 2) just pass everything back and let the user deal with it in the `getCurrentUser` function?

Reply:
@jtoar In my experience very little actually comes back in the `jwtPayload` unless you have customized a jwt template or the session token in clerk. That said I just tested this locally by updating the session token in clerk with the following:

```json
{
  "publicMetadata": "{{user.public_metadata}}",
  "id": "{{user.id}}",
  "emailAddress": "{{user.primary_email_address}}"
}
```

I changed the auto decoder slightly to return this:

```javascript
const roles = jwtPayload?.publicMetadata?.['roles'] ?? []

return {
  ...jwtPayload,
  roles,
  id: jwtPayload.sub,
}
```

For my app's purpose this worked. This is probably more of a documentation issue. For a lot of people, just making sure the `id: jwtPayload.sub` is returned is probably enough and they can deal with it in `getCurrentUser` as you suggested as long as the entire `jwtPayload` is returned. In theory we wouldn't even need to return roles here because we could get that based on how we defined it in Clerk from within the `getCurrentUser` function as well.

Reply:
Thanks @o0charlie0o this sounds like the way to go, let's keep it simple. `roles` could come from the user's database instead of Clerk—we just don't know and probably shouldn't be trying to figure it out at the decoder level.

Reply:
I removed this line and the roles return in https://github.com/redwoodjs/redwood/pull/8642/commits/e6a8edc8b08ef5de0cd37cd4b9f225af07435117